src/common/webrequest_curl.cpp

```diff
@@ -0,0 +1,559 @@
+    // Set callback functions
+    curl_easy_setopt(m_handle, CURLOPT_WRITEFUNCTION, wxCURLWriteData);
+    curl_easy_setopt(m_handle, CURLOPT_HEADERFUNCTION, wxCURLHeader);
+    curl_easy_setopt(m_handle, CURLOPT_READFUNCTION, wxCURLRead);
+    curl_easy_setopt(m_handle, CURLOPT_READDATA, static_cast<void*>(this));
+    // Enable gzip, etc decompression
+    curl_easy_setopt(m_handle, CURLOPT_ACCEPT_ENCODING, "");
+    // Enable redirection handling
+    curl_easy_setopt(m_handle, CURLOPT_FOLLOWLOCATION, 1L);
+    // Limit redirect to HTTP
+    curl_easy_setopt(m_handle, CURLOPT_REDIR_PROTOCOLS,
+        CURLPROTO_HTTP | CURLPROTO_HTTPS);
+    // Enable all supported authentication methods
+    curl_easy_setopt(m_handle, CURLOPT_HTTPAUTH, CURLAUTH_ANY);
+    curl_easy_setopt(m_handle, CURLOPT_PROXYAUTH, CURLAUTH_ANY);
+}
+
+wxWebRequestCURL::~wxWebRequestCURL()
+{
+    DestroyHeaderList();
+
+    curl_easy_cleanup(m_handle);
+}
+
+void wxWebRequestCURL::Start()
+{
+    if ( GetState() != State_Idle )
+        return;
+
+    m_response.reset(new wxWebResponseCURL(*this));
+    m_response->Init();
+
+    if ( m_dataSize )
+    {
+        if ( m_method.empty() || m_method.IsSameAs("POST", false) )
+        {
+            curl_easy_setopt(m_handle, CURLOPT_POSTFIELDSIZE_LARGE,
+                static_cast<curl_off_t>(m_dataSize));
+            curl_easy_setopt(m_handle, CURLOPT_POST, 1L);
+        }
+        else if ( m_method.IsSameAs("PUT", false) )
+        {
+            curl_easy_setopt(m_handle, CURLOPT_UPLOAD, 1L);
+            curl_easy_setopt(m_handle, CURLOPT_INFILESIZE_LARGE,
+                static_cast<curl_off_t>(m_dataSize));
+        }
+    }
+
+    if ( m_method.IsSameAs("HEAD", false) )
+    {
+        curl_easy_setopt(m_handle, CURLOPT_NOBODY, 1L);
+    }
+    else if ( !m_method.empty() )
+    {
+        curl_easy_setopt(m_handle, CURLOPT_CUSTOMREQUEST,
+            static_cast<const char*>(m_method.mb_str()));
+    }
+
+    for ( wxWebRequestHeaderMap::const_iterator it = m_headers.begin();
+        it != m_headers.end(); ++it )
+    {
+        wxString hdrStr = wxString::Format("%s: %s", it->first, it->second);
+        m_headerList = curl_slist_append(m_headerList, hdrStr.mb_str());
+    }
+    curl_easy_setopt(m_handle, CURLOPT_HTTPHEADER, m_headerList);
+
+    StartRequest();
+}
+
+bool wxWebRequestCURL::StartRequest()
+{
+    m_bytesSent = 0;
+
+    if ( static_cast<wxWebSessionCURL&>(GetSession()).StartRequest(*this) )
+    {
+        SetState(State_Active);
+        return true;
+    }
+    else
+    {
+        SetState(State_Failed);
+        return false;
+    }
+}
+
+void wxWebRequestCURL::Cancel()
+{
+    static_cast<wxWebSessionCURL&>(GetSession()).CancelRequest(this);
+}
+
+void wxWebRequestCURL::HandleCompletion()
+{
+    int status = (m_response) ? m_response->GetStatus() : 0;
+
+    if ( status == 0)
+        SetState(State_Failed, GetError());
+    else if ( status == 401 || status == 407 )
```

Comment:
I know we all know HTTP status codes by heart (it surprised me that I did remember that 407 was "proxy auth required", let alone what 401 was...), but it would still be nice to define symbolic constants for them IMO.